x/qgb/abci.go

```diff
@@ -113,3 +121,89 @@ func handleValsetRequest(ctx sdk.Context, k keeper.Keeper) {
 		}
 	}
 }
+
+// PruneIfNeeded runs basic checks on saved attestations to see if we need to prune or not.
+// Also, proposes some gas optimizations by adding more checks to return before running any
+// of the pruning logic.
+func PruneIfNeeded(ctx sdk.Context, k keeper.Keeper) {
+	// If the attestations nonce hasn't been initialized yet, no pruning is
+	// required
+	if !k.CheckLatestAttestationNonce(ctx) {
+		return
+	}
+	// If the nonce is not greater than the minimum number of attestations,
+	// no pruning is required.
+	if k.GetLatestAttestationNonce(ctx) <= AttestationPruningThreshold {
+		return
+	}
+	lastAvailableNonce := k.GetLastAvailableAttestationNonce(ctx)
+	lastUnbondingNonce := k.GetLastUnbondingNonce(ctx)
+	if lastAvailableNonce == lastUnbondingNonce {
+		// we don't need to do anything, we want to keep attestations up to the last unbonding nonce
+		return
+	}
+	if lastAvailableNonce > lastUnbondingNonce {
+		// checking if it's the initial case following the startup of the chain
+		if lastAvailableNonce == 1 && lastUnbondingNonce == 0 {
+			return
+		}
+		// we should never hit this case, since we will keep the attestations up to the last unbonding nonce
+		ctx.Logger().Error("missing attestations up to the unbonding nonce")
+		return
+	}
+	// now we have attestations before the unbonding height, we should check whether we need to prune them
+	// or not yet
+	latestAttestationNonce := k.GetLatestAttestationNonce(ctx)
+	if latestAttestationNonce-lastAvailableNonce+1 <= AttestationPruningThreshold {
+		// we don't need to prune as we still have room to store more attestations
+		return
+	}
+	// now we want to prune attestations as we have the following conditions:
+	// - we have attestations up to the last unbonding height
+	// - the total number of attestations, including the ones before the unbonding height, are greater
+	// than the AttestationPruningThreshold
+	err := pruneAttestations(ctx, k, lastAvailableNonce, latestAttestationNonce, lastUnbondingNonce)
+	if err != nil {
+		ctx.Logger().Error("error while pruning QGB attestations", "error", err.Error())
+		return
+	}
+}
+
+// pruneAttestations prunes attestations to keep attestations up to the last unbonding height
+// and the total number of attestations lower than the AttestationPruningThreshold.
+func pruneAttestations(
+	ctx sdk.Context,
+	k keeper.Keeper,
+	lastAvailableAttestationNonce uint64,
+	latestAttestationNonce uint64,
+	lastUnbondingNonce uint64,
+) error {
+	ctx.Logger().Debug("pruning attestations from qgb store")
```

Comment:
[uber nit] for consistency with the log line above.

```go
		ctx.Logger().Error("error while pruning QGB attestations", "error", err.Error())
```

```suggestion
	ctx.Logger().Debug("pruning attestations from QGB store")
```